tests/test_slips_utils.py

```diff
@@ -1,10 +1,410 @@
 from tests.module_factory import ModuleFactory
+import datetime
+import ipaddress
+from unittest.mock import patch
+import requests
+from dataclasses import dataclass
+from enum import Enum
+import sys
+import pytest
+import socket
+
+
+class MockFlow:
+    def __init__(self, proto, starttime, saddr, daddr, sport, dport):
+        self.proto = proto
+        self.starttime = starttime
+        self.saddr = saddr
+        self.daddr = daddr
+        self.sport = sport
+        self.dport = dport
 
 
 def test_get_hash_from_file():
     utils = ModuleFactory().create_utils_obj()
     # a file that we know doesn't change
     assert (
-        utils.get_hash_from_file("modules/template/__init__.py")
+        utils.get_sha256_hash("modules/template/__init__.py")
         == "2d12747a3369505a4d3b722a0422f8ffc8af5514355cdb0eb18178ea7071b8d0"
     )
+
+
+def test_get_hash_from_nonexistent_file():
+    utils = ModuleFactory().create_utils_obj()
+    with pytest.raises(FileNotFoundError):
+        utils.get_sha256_hash("nonexistent_file.txt")
+
+
+def test_get_hash_from_file_permission_error():
+    utils = ModuleFactory().create_utils_obj()
+    with patch("builtins.open", side_effect=PermissionError):
+        with pytest.raises(PermissionError):
+            utils.get_sha256_hash("restricted_file.txt")
+
+
+@pytest.mark.parametrize(
+    "input_string, expected_output",
+    [
+        ("Hello; world `& |$(this", "Helloworld`this"),  # special chars
+        ("", ""),  # empty input
+        ("!@#$%^&*()", ""),  # input with only special characters
+        ("Thisisacleanstring", "Thisisacleanstring"),
+        ("Hello World!", "HelloWorld"),  # input with spaces
+    ],
+)
+def test_sanitize(input_string, expected_output):
+    utils = ModuleFactory().create_utils_obj()
+    assert utils.sanitize(input_string) == expected_output
+
+
+@pytest.mark.parametrize(
+    "data, expected_type",
+    [
+        ("192.168.1.100", "ip"),
+        ("2001:0db8:85a3:0000:0000:8a2e:0370:7334", "ip"),
+        ("192.168.0.0/16", "ip_range"),
+        ("e10adc3949ba59abbe56e057f20f883e", "md5"),
+        ("example.com", "domain"),
+        ("http://example.com/some/path", "url"),
+        ("AS12345", "asn"),
+        ("999.999.999.999", None),
+        ("123456789abcdefg", None),
+    ],
+)
+def test_detect_data_type(data, expected_type):
+    utils = ModuleFactory().create_utils_obj()
+    assert utils.detect_data_type(data) == expected_type
+
+
+def test_get_first_octet():
```

Comment:
mark.parametrize can be used to improve this function

Reply:
hey @Sekhar-Kumar-Dash  this comment wasnt resolved right? you didn't use mark.parametrize in test_get_first_octet()